### PR TITLE
Start and Stop signal registration for the FlaggedStorage

### PR DESCRIPTION
**Pull request by AndreaCatania**
Sometimes you may want to perform some operations on the storage, but you don't want
that these operations produce any event.

You can use the function `storage.set_event_emission(false)` to suppress the event writing for
of any action. When you want to re activate them you can simply call
`storage.set_event_emission(true)`

This change will be used to fix this: https://github.com/amethyst/amethyst/issues/1795

## Checklist

* [x] I've added tests for all code changes and additions (where applicable)
* [x] I've added a demonstration of the new feature to one or more examples
* [x] I've updated the book to reflect my changes
* [x] Usage of new public items is shown in the API docs

### CODE REVIEW

**Comment by jaynus** [comment]
Wouldn't setting `storage.set_event_emission(false)` accidentally allow for https://github.com/slide-rs/specs/issues/541 in some cases?

**Comment by AndreaCatania** [comment]
Do you mean that when the emission is false, the parallel joint is allowed?

I think that this change doesn't fix or make worse that scenario, because the main point of that issue
is find a way to join in parallel the flagged storages and write the events.

So I would not consider this as a fix or a partial fix.

**Comment by jojolepro** [comment]
bors try

**Comment by torkleyy** [comment]
> Sometimes you may want to perform some operations on the storage, but you don't want
that these operations produce any event.

Why do you want that? It seems to me in this case the requirement stems from unclean design. `FlaggedStorage` is just meant to be a very simple wrapper which just gives you all the events.

If you want a more custom storage, please consider making your own instead of adding it to Specs; I'm trying to keep Specs small, and stabilize it.

**Comment by AndreaCatania** [comment]
The reason is that sometimes a particular `System` perform some operations on the storage and these don't have to be propagated.
Like the parenting hierarchy resolution of the `Transform` component in `Amethyst` here https://github.com/amethyst/amethyst/issues/1795 .

Create a new Storage to solve this problem is doable, but mean also copy an paste 80% of the actual `FlaggedStorage`.

If you and the others are ok with it, I can create a new storage type that is capable of start and stop event signaling and so solve that issue.

**Comment by AndreaCatania** [comment]
@torkleyy However this functionality is not so invasive, and only add a feature that can be useful even out of Amethyst.

Is always a good thing be able to `start` and `stop` event submission.

**Comment by AndreaCatania** [comment]
Hello guys! Any idea about this PR?

**Comment by jojolepro** [comment]
I'm okay with merging this, but if @torkleyy doesn't want to, then it would make sense to create this type of storage in an external crate and use it in amethyst.

The main downside is that using an external crate will require more maintenance.

**Comment by AndreaCatania** [comment]
I waited a bit to hear the @torkleyy thoughts on this;
I assume that he don't want to merge it, so I'll create a new crate in the next few days.

_(By the way, I think that use a different crate for this is not the optimal solution in this case)_

**Comment by AndreaCatania** [comment]
Yes I can to it.

> I think it would be fine to merge it when it's disabled by default; 

Is it already disabled by default.

> Can we put this under a feature flag?

Would you like to use a flag (i32 - bit wise operations) instead of a boolean?

**Comment by AndreaCatania** [comment]
Re-thinking about your comment, probably your idea is to use the conditional building using the features. Let me know which one

**Comment by jojolepro** [comment]
feature flag = "conditional building" = `features = ["somethingsomething"]`

**Comment by AndreaCatania** [comment]
@jojolepro Thanks for the clarification!
@torkleyy Done!

File: src/storage/flagged.rs [hunk 1]
```diff
@@ -166,16 +201,37 @@ impl<C: Component, T: UnprotectedStorage<C>> UnprotectedStorage<C> for FlaggedSt
...
+                self.channel.single_write(ComponentEvent::Inserted(id));
+            }
+        }
+        #[cfg(not(feature = "storage-event-control"))]
         self.channel.single_write(ComponentEvent::Inserted(id));
         self.storage.insert(id, comp);
     }
 
     unsafe fn remove(&mut self, id: Index) -> C {
+        #[cfg(feature = "storage-event-control")]
```
Comment (torkleyy) [inline src/storage/flagged.rs]: Can you please define a function which just returns `true` in case of a disabled feature?

**Comment by AndreaCatania** [comment]
@torkleyy Done, now looks much better.

**Comment by AndreaCatania** [comment]
Morning, any idea about this?

**Pull request merged**